proposals/0391-replace-stake-program-floating-point.md

```diff
@@ -0,0 +1,138 @@
+preserve precision while using integer arithmetic.
+
+- The floating-point rates will be converted to their fractional
+  equivalents:
+  - `DEFAULT_WARMUP_COOLDOWN_RATE` (`0.25`) becomes a fraction of
+    **(numerator: 1, denominator: 4)**.
+  - `NEW_WARMUP_COOLDOWN_RATE` (`0.09`) becomes a fraction of
+    **(numerator: 9, denominator: 100)**.
+
+- The current flow
+  `(account_portion / cluster_portion) * (cluster_effective * rate)` is
+  reordered to
+  `(account_portion * cluster_effective * rate_numerator) / (cluster_portion * rate_denominator)`.
+  Instead of performing divisions early in the process (which truncates
+  intermediate results), all multiplications are performed first.
+
+- All intermediate multiplications are performed using `u128` to
+  prevent overflow.
+
+### State Compatibility
+
+To maintain backwards compatibility with on-chain stake account data,
+the `Delegation` struct will be modified:
+
+```diff
+pub struct Delegation {
+    pub voter_pubkey: Pubkey,
+    pub stake: u64,
+    pub activation_epoch: Epoch,
+    pub deactivation_epoch: Epoch,
+-   pub warmup_cooldown_rate: f64,
++    _reserved: [u8; 8],
+}
+```
+
+This preserves the exact memory size and layout of existing accounts.
+It is a legacy field anyway, with the actual rate being determined
+dynamically in functions.
+
+## Alternatives Considered
+
+Decimal scaling factor. Uses a uniform scaling factor represent rates
+as integers like BPS (e.g 0.25 = 2500). Rejected as rational
+arithmetic is more precise. It may be easier to use for external
+consumers, but these values are really only used internally to the
+stake interface crate.
+
+## Impact
+
+- **Stake Program:** The program must be updated to use the new
+  integer-based calculation helpers from `solana-stake-interface`. It is
+  doing so mostly through its use of
+  `stake.delegation.stake_activating_and_deactivating()`. Also, the new
+  `Delegation` struct definition (with its private field), will likely
+  impact how its being instantiated in a few areas.
+
+- **Agave:** Update the workspace dependency on
+  `solana-stake-interface`. Runtime stake processing already funnels
+  through Delegation::stake_activating_and_deactivating, so a dependency
+  bump automatically picks up the fixed-point math without touching
+  Agave code.
+
+- **Firedancer:** Will also need to update their stake interface
+  dependency in lockstep with Agave.
+
+## Security Considerations
+
+1. **Unit tests:** Baseline of correctness by testing specific, known
+   scenarios and edge cases.
+2. **Differential Fuzzing (`proptest`):**
+   - An oracle function preserving the original `f64` logic will be
+     maintained for testing purposes only.
+   - The test will run the new integer implementation against the oracle
+     with millions of random inputs.
+   - Assert that the results are within a relative tolerance to account
+     for the increased precision.
```

Comment:
This is a bit vague. What is this tolerance supposed to be?

The exact difference will depend on the size of the stake and the amount on mainnet-beta, or even better, give examples of how different the calculation can be for small and large stake accounts.

But since we're essentially widening out the calculation to 128 bits instead of 64, the new calculation should only diverge past ~52 bits?

Either way, since the calculation is more accurate, this should actually give people *higher* amounts of activation / deactivation, correct?